src/main/java/com/shapesecurity/shift/es2016/template/Template.java

```diff
@@ -0,0 +1,207 @@
+				throw new IllegalArgumentException("Unrecognized node type \"" + typeName + "\"");
+			}
+		}
+
+		return Maybe.of(Pair.of(name, Maybe.empty()));
+	}
+
+	public static ImmutableList<NodeInfo> findNodes(Program tree, WithLocation locations, ImmutableList<ParserWithLocation.Comment> comments) {
+		// TODO: allow custom matcher
+		ImmutableList<Marker> markers = Maybe.catMaybes(
+			comments.map(c -> parseComment(c.text).map(p -> new Marker(p.left, p.right, c.start.offset, c)))
+		);
+
+		if (markers.isEmpty()) {
+			return ImmutableList.empty();
+		}
+
+		ImmutableList<Node> nodes = Flattener.flatten(tree);
+		ImmutableList<Pair<Node, SourceSpan>> nodeAndLocations = Maybe.catMaybes(nodes.map(node -> {
+			Maybe<SourceSpan> maybeLocation = locations.getLocation(node);
+			if (maybeLocation.isNothing()) {
+				if (node instanceof BindingIdentifier && ((BindingIdentifier) node).name.equals("*default*")) {
+					return Maybe.empty();
+				}
+				throw new IllegalArgumentException("Missing location information for node " + node); // TODO there is possibly a better type to throw
+			}
+			SourceSpan location = maybeLocation.fromJust();
+			return Maybe.of(Pair.of(node, location));
+		}));
+		@SuppressWarnings("unchecked")
+		Pair<Node, SourceSpan>[] nodeArray = nodeAndLocations.toArray(new Pair[nodeAndLocations.length]);
+		Arrays.sort(nodeArray, Comparator.comparing(p -> p.right.start.offset));
+
+		LinkedList<Pair<Integer, LinkedList<Pair<Node, Integer>>>> joinedByStart = new LinkedList<>();
+		joinedByStart.add(Pair.of(0, new LinkedList<>()));
+		for (Pair<Node, SourceSpan> nodeAndLocation : nodeArray) {
+			int offset = nodeAndLocation.right.start.offset;
+			if (joinedByStart.getLast().left < offset) {
+				joinedByStart.add(Pair.of(offset, new LinkedList<>()));
+			}
+			assert joinedByStart.getLast().left.equals(offset);
+			joinedByStart.getLast().right.add(Pair.of(nodeAndLocation.left, nodeAndLocation.right.end.offset));
+		}
+
+		final ArrayList<NodeInfo> out = new ArrayList<>(markers.length);
+		Iterator<Pair<Integer, LinkedList<Pair<Node, Integer>>>> nodeWalker = joinedByStart.iterator();
+		assert nodeWalker.hasNext();
+		Pair<Integer, LinkedList<Pair<Node, Integer>>> currentBlockOfNodes = nodeWalker.next();
+		for (Marker marker : markers) {
+			while (currentBlockOfNodes.left < marker.start) {
+				if (!nodeWalker.hasNext()) {
+					throw new IllegalArgumentException("Couldn't find node following marker " + marker.name);
+				}
+				currentBlockOfNodes = nodeWalker.next();
+			}
+
+			Stream<Pair<Node, Integer>> nodeStream = currentBlockOfNodes.right.stream();
+			List<Pair<Node, Integer>> ofCorrectType = marker.type.map(type -> nodeStream.filter(p -> type.isInstance(p.left))).orJust(nodeStream).collect(Collectors.toList());
+			if (ofCorrectType.isEmpty()) {
+				throw new IllegalArgumentException("Couldn't find any nodes of type " + marker.type.fromJust().getSimpleName() + " for marker " + marker.name);
```

Comment:
Do we want it to throw in this case or just not be present in the output? I'm not sure.

Reply:
I strongly favor throwing here. If this happens the template is ill-formed, at least with respect to the particular choice of comment-parsing functions.